zookeeper/source/zk_nn_client.cc

```diff
@@ -21,359 +22,359 @@
-	if (zk->get_children(BLOCK_LOCATIONS + block_id_str, children, error_code)) {
-	    if (children.size() >= MIN_REPLICATION){
-		return true;						
-	    }
-	    return false;
+	using namespace hadoop::hdfs;
+
+	const std::string ZkNnClient::CLASS_NAME = ": **ZkNnClient** : ";
+
+	ZkNnClient::ZkNnClient(std::string zkIpAndAddress) : ZkClientCommon(zkIpAndAddress) {
+		mkdir_helper( "/", false);
+	}
+
+	ZkNnClient::ZkNnClient(std::shared_ptr <ZKWrapper> zk_in) : ZkClientCommon(zk_in) {
+		mkdir_helper( "/", false);
+	}
+
+	/*
+	 * A simple print function that will be triggered when
+	 * namenode loses a heartbeat
+	 */
+	void notify_delete() {
+		printf("No heartbeat, no childs to retrieve\n");
+	}
+
+	void ZkNnClient::register_watches() {
+
+		int error_code;
+		std::vector <std::string> children = std::vector <std::string>();
+
+		/* Place a watch on the health subtree */
+
+
+		if (!(zk->wget_children(HEALTH, children, zk->watcher_health_factory(HEALTH_BACKSLASH), nullptr, error_code))) {
+			// TODO: Handle error
+			LOG(ERROR) << CLASS_NAME << "[In register_watchers], wget failed " << error_code;
+		}
+
+		for (int i = 0; i < children.size(); i++) {
+			LOG(INFO) << CLASS_NAME << "[In register_watches] Attaching child to " << children[i] << ", ";
+			std::vector <std::string> ephem = std::vector <std::string>();
+			if(!(zk->wget_children(HEALTH_BACKSLASH + children[i], ephem, zk->watcher_health_child, nullptr, error_code))) {
+				// TODO: Handle error
+				LOG(ERROR) << CLASS_NAME << "[In register_watchers], wget failed " << error_code;
+			}
+			/*
+			   if (ephem.size() > 0) {
+			   LOG(INFO) << CLASS_NAME << "Found ephem " << ephem[0];
+			   } else {
+			   LOG(INFO) << CLASS_NAME << "No ephem found for " << children[i];
+			   }
+			 */
+		}
+	}
+
+	bool ZkNnClient::file_exists(const std::string& path) {
+		bool exists;
+		int error_code;
+		if (zk->exists(ZookeeperPath(path), exists, error_code)) {
+			return exists;
+		} else {
+			// TODO: Handle error
+		}
+	}
+
+
+
+	// --------------------------- PROTOCOL CALLS ---------------------------------------
+
+	void ZkNnClient::read_file_znode(FileZNode& znode_data, const std::string& path) {
+		int error_code;
+		std::vector<std::uint8_t> data(sizeof(znode_data));
+		if (!zk->get(ZookeeperPath(path), data, error_code)) {
+			LOG(ERROR) << "We could not read the file znode at " << path;
+			return; // don't bother reading the data
+		}
+		std::uint8_t *buffer = &data[0];
+		memcpy(&znode_data, buffer, sizeof(znode_data));
+	}
+
+	void ZkNnClient::file_znode_struct_to_vec(FileZNode* znode_data, std::vector<std::uint8_t>& data) {
+		memcpy(&data[0], znode_data, sizeof(*znode_data));
 	}
-	return true;
-    }
 
-    bool ZkNnClient::add_block(AddBlockRequestProto& req, AddBlockResponseProto& res) {
+
+	bool ZkNnClient::previousBlockComplete(uint64_t prev_id) {
+		int error_code;
+		/* this value will eventually be read from config file */
+		int MIN_REPLICATION = 1;
```

Comment:
I believe this is also on a per-file basis. It is a field in a file znode. So we would need to read the file that the block belongs to...is it even possible to get a blockid -> file mapping?